crates/fiber-lib/src/store/pub_sub/store_with_pub_sub.rs

```diff
@@ -0,0 +1,286 @@
+    }
+
+    fn insert_invoice(
+        &self,
+        invoice: CkbInvoice,
+        preimage: Option<Hash256>,
+    ) -> Result<(), InvoiceError> {
+        let invoice_hash = *invoice.payment_hash();
+        self.inner.insert_invoice(invoice, preimage)?;
+
+        self.publish(StoreUpdatedEvent::new_invoice_updated_event(
+            invoice_hash,
+            InvoiceUpdatedPayload::Open,
+        ));
+        Ok(())
+    }
+
+    fn update_invoice_status(
+        &self,
+        id: &Hash256,
+        status: CkbInvoiceStatus,
+    ) -> Result<(), InvoiceError> {
+        let _span = tracing::info_span!("update_invoice_status", invoice_hash = ?id).entered();
+
+        self.inner.update_invoice_status(id, status)?;
+        let payload_opt = status.try_into().ok().or_else(|| match status {
+            CkbInvoiceStatus::Received => {
+                // TODO: We should save received amount to the store. This is useful to
+                // determine if the invoice is fully paid.
+                // Currently we assume that the invoice is fully paid if the status is
+                // Received.
+                let payload_opt =
+                    self.get_invoice(id)
+                        .and_then(|invoice| invoice.amount)
+                        .map(|amount| InvoiceUpdatedPayload::Received {
+                            amount,
+                            is_finished: true,
+                        });
+                if payload_opt.is_none() {
+                    tracing::error!("Fail to get payment amount for a received invoice");
+                }
+                payload_opt
+            }
+            _ => {
+                tracing::error!(
+                    "Expect convert CkbInvoiceStatus {} to InvoiceUpdatedPayload",
+                    status
+                );
+                None
+            }
+        });
+        if let Some(payload) = payload_opt {
+            self.publish(StoreUpdatedEvent::new_invoice_updated_event(*id, payload))
+        }
+        Ok(())
+    }
+
+    fn get_invoice_status(&self, id: &Hash256) -> Option<CkbInvoiceStatus> {
+        self.inner.get_invoice_status(id)
+    }
+}
+
+// The PaymentUpdatedEvent requires preimage when the payment status is Received. Hooks are added
+// in both `insert_preimage` and `insert_payment_session` so App does not need to worry about
+// sequence to update the store.
+impl<S> StoreWithPubSub<S>
+where
+    S: NetworkGraphStateStore + PreimageStore,
+{
+    fn publish_payment_updated_event_when_inserting_payment_session(
+        &self,
+        payment_hash: Hash256,
+        status: PaymentStatus,
+    ) {
+        let payload_opt = status.try_into().ok().or_else(|| match status {
+            // If preimage is not available, defer the notification on `insert_preimage`.
+            // See `publish_payment_updated_event_when_removing_preimage`.
+            PaymentStatus::Success => self
+                .get_preimage(&payment_hash)
```

Comment:
please note we only call insert_preimage for a payment for middle hops of a payment path,
so it only make sense to call get_preimage for these hops,
and we only have payment session in the payment sender.

so i'm not sure how can we get_preimage when inserting payment session?